### PR TITLE
Add support for column level access control

### PR DESCRIPTION
**Pull request by rschlussel2**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 1]
```diff
@@ -683,6 +685,36 @@ protected Scope visitQuery(Query node, Optional<Scope> scope)
...
                     .withRelationType(RelationId.of(node), queryBodyScope.getRelationType())
                     .build();
+
+            // create a multimap of tables in a query to columns used for each table
+            Multimap<QualifiedObjectName, String> tableColumnMap = analysis.getColumnReferenceFields().values().stream()
+                    .filter(fieldId -> {
+                        // first filter out column references that don't directly reference tables
+                        Node sourceNode = fieldId.getRelationId().getSourceNode();
+                        return sourceNode instanceof Table
+                                && analysis.getScope(sourceNode)
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: Why do you need this?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 2]
```diff
@@ -683,6 +685,36 @@ protected Scope visitQuery(Query node, Optional<Scope> scope)
                     .withParent(withScope)
                     .withRelationType(RelationId.of(node), queryBodyScope.getRelationType())
                     .build();
+
+            // create a multimap of tables in a query to columns used for each table
+            Multimap<QualifiedObjectName, String> tableColumnMap = analysis.getColumnReferenceFields().values().stream()
+                    .filter(fieldId -> {
+                        // first filter out column references that don't directly reference tables
+                        Node sourceNode = fieldId.getRelationId().getSourceNode();
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: This looks like a hack. I would prefer to collect this information more explicitly in `Analysis`, just like we do `com.facebook.presto.sql.analyzer.Analysis#tables`. Or find all `Table` nodes in the AST...

This brings an interesting question though. Assuming I don't have right to select from `tab_b.d`, should the following query fail?
```
select * from tab_a where exists (select d from tab_b); -- column `d` can be pruned
```

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 3]
```diff
@@ -683,6 +685,36 @@ protected Scope visitQuery(Query node, Optional<Scope> scope)
...
+                                            .getOriginTable().orElseThrow(() -> new IllegalStateException("No origin table for field")),
+                                    // column name
+                                    fieldId -> analysis.getScope(fieldId.getRelationId().getSourceNode())
+                                            .getRelationType()
+                                            .getFieldByIndex(fieldId.getFieldIndex())
+                                            .getName().orElseThrow(() -> new IllegalStateException("No name for field")),
+                                    ArrayListMultimap::create));
+
+            // check column access permissions for each table
+            tableColumnMap.keySet()
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: `tableColumnMap.forEach`

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/RelationId.java [hunk 4]
```diff
@@ -40,6 +40,12 @@ public static RelationId anonymous()
         return new RelationId(null);
     }
 
+    @Nullable
+    public Node getSourceNode()
+    {
+        return sourceNode;
+    }
+
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/RelationId.java]: it would be nice to separate this changes to different commit. Analyzer is completely different than engine security plugins.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 5]
```diff
@@ -683,6 +685,36 @@ protected Scope visitQuery(Query node, Optional<Scope> scope)
                     .withParent(withScope)
                     .withRelationType(RelationId.of(node), queryBodyScope.getRelationType())
                     .build();
+
+            // create a multimap of tables in a query to columns used for each table
+            Multimap<QualifiedObjectName, String> tableColumnMap = analysis.getColumnReferenceFields().values().stream()
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: wrap this as method

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 2 above)
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: >  I would prefer to collect this information more explicitly in Analysis, just like we do `com.facebook.presto.sql.analyzer.Analysis#tables`

Sounds good to me.

File: presto-plugin-toolkit/src/main/java/com/facebook/presto/plugin/base/security/AllowAllAccessControl.java [hunk 6]
```diff
@@ -25,6 +25,21 @@
 public class AllowAllAccessControl
         implements ConnectorAccessControl
 {
+    @Override
```
Comment (kokosing) [inline presto-plugin-toolkit/src/main/java/com/facebook/presto/plugin/base/security/AllowAllAccessControl.java]: can you add a test to prevent such situation in future

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 7]
```diff
@@ -683,6 +685,36 @@ protected Scope visitQuery(Query node, Optional<Scope> scope)
...
+                        return sourceNode instanceof Table
+                                && analysis.getScope(sourceNode)
+                                .getRelationType()
+                                .getFieldByIndex(fieldId.getFieldIndex())
+                                .getOriginTable().isPresent();
+                    })
+                    .collect(
+                            toMultimap(
+                                    // table associated with the column
+                                    fieldId -> analysis.getScope(fieldId.getRelationId().getSourceNode())
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: Getting scope for relation id is a temporary solution for the planner to work. In the future, this mapping will be gone, so please don't rely on it.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 2 above)
Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: You could gather references to table fields via `ExpressionAnalyzer.Visitor#handleResolvedField`. You should be always able to access `Field` there.

File: presto-plugin-toolkit/src/main/java/com/facebook/presto/plugin/base/security/AllowAllAccessControl.java (hunk 6 above)
Comment (findepi) [reply presto-plugin-toolkit/src/main/java/com/facebook/presto/plugin/base/security/AllowAllAccessControl.java]: Shall we generalize `TestPartitionsAwareAccessControl` to other connectors?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 2 above)
Comment (rschlussel2) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: @findepi  Thanks for the suggestion.  That's a much better idea.  Also, I don't think we should worry about columns that can be pruned.  If a user doesn't want us checking access for a column, they shouldn't put it in the query.  In fact, it might be confusing for it not to fail if it fails for other queries.

Comment (martint) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: Pruning is an optimization. Semantically, the columns are read, so the query should fail. And that’s what the spec says, btw.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 1 above)
Comment (rschlussel2) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: for queries like `WITH a AS (SELECT 1) (SELECT * FROM a);`  you end up with a `sourceNode` of type `TABLE` for the `SELECT * FROM a`, but it doesn't have an origin table.

File: presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorAccessControl.java [hunk 8]
```diff
@@ -287,4 +289,14 @@ default void checkCanRevokeTablePrivilege(ConnectorTransactionHandle transaction
...
     }
+
+    /**
+     * Check if identity is allowed to select from the specified columns in a table.
+     *
+     * @throws com.facebook.presto.spi.security.AccessDeniedException if not allowed
+     */
+    default void checkCanSelectFromColumns(ConnectorTransactionHandle transactionHandle, Identity identity, SchemaTableName tableName, Collection<String> columnNames)
+    {
+        denySelectColumns(tableName.toString(), columnNames);
```
[the remark this answers is not in the document]
Comment (rschlussel2) [reply presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorAccessControl.java]: I was also worried about this, but it's been true for every kind of access control check we've added.  I can't think of a better approach.

File: presto-plugin-toolkit/src/main/java/com/facebook/presto/plugin/base/security/AllowAllAccessControl.java (hunk 6 above)
Comment (rschlussel2) [reply presto-plugin-toolkit/src/main/java/com/facebook/presto/plugin/base/security/AllowAllAccessControl.java]: I'll do that as a follow on.  This was just a simple opportunistic fix, and I don't want to get caught up with it in this PR.

Comment (findepi) [reply presto-plugin-toolkit/src/main/java/com/facebook/presto/plugin/base/security/AllowAllAccessControl.java]: What about extracting `Add create/delete/rename schema in the connector AllowAllAccessControl` to a separate PR then?

File: presto-hive/src/main/java/com/facebook/presto/hive/security/SqlStandardAccessControl.java [hunk 9]
```diff
@@ -281,6 +282,12 @@ public void checkCanRevokeTablePrivilege(ConnectorTransactionHandle transaction,
         }
     }
 
+    @Override
+    public void checkCanSelectFromColumns(ConnectorTransactionHandle transactionHandle, Identity identity, SchemaTableName tableName, Collection<String> columnNames)
+    {
+        // Allow: No column-level privileges in Hive
```
[the remark this answers is not in the document]
Comment (rschlussel2) [reply presto-hive/src/main/java/com/facebook/presto/hive/security/SqlStandardAccessControl.java]: for now I added todos.  I'll get to this soon.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java [hunk 10]
```diff
@@ -386,6 +395,11 @@ private Type handleResolvedField(Expression node, FieldId fieldId, Type resolved
                 }
             }
 
+            Field field = baseScope.getRelationType().getFieldByIndex(fieldId.getFieldIndex());
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java]: I think it would be better to modify `handleResolvedField(.., ResolvedField, ..)`, since then you have direct access to `resolvedField.getField()`

File: presto-plugin-toolkit/src/main/java/com/facebook/presto/plugin/base/security/ReadOnlyAccessControl.java [hunk 11]
```diff
@@ -144,6 +143,12 @@ public void checkCanCreateViewWithSelectFromView(ConnectorTransactionHandle tran
         // allow
     }
 
+    @Override
+    public void checkCanCreateViewWithSelectFromColumns(ConnectorTransactionHandle transactionHandle, Identity identity, SchemaTableName tableName, Set<String> columnNames)
```
Comment (findepi) [inline presto-plugin-toolkit/src/main/java/com/facebook/presto/plugin/base/security/ReadOnlyAccessControl.java]: Shouldn't `checkCanCreateViewWithSelectFromColumns` go always right after `checkCanSelectFromColumns`?

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 12]
```diff
@@ -7391,7 +7391,10 @@ public void testAccessControl()
...
-        assertAccessDenied("SELECT nationkey, regionkey from nation", "Cannot select from columns .nationkey, regionkey. in table .*.nation.*", privilege("nationkey", SELECT_COLUMN));
+        assertAccessDenied("SELECT n1.nationkey, n2.regionkey FROM nation n1, nation n2", "Cannot select from columns .nationkey, regionkey. in table .*.nation.*", privilege("nationkey", SELECT_COLUMN));
+        assertAccessDenied("SELECT count(name) as c FROM nation where comment > 'abc' GROUP BY regionkey having max(nationkey) > 10", "Cannot select from columns .name, nationkey, comment, regionkey. in table .*.nation.*", privilege("nationkey", SELECT_COLUMN));
+        assertAccessDenied("SELECT nationkey FROM (SELECT * FROM nation)", "Cannot select from columns .name, nationkey, comment, regionkey. in table .*.nation.*", privilege("nationkey", SELECT_COLUMN));
+        assertAccessDenied("SELECT 1 FROM region, nation where region.regionkey = nation.nationkey", "Cannot select from columns .nationkey. in table .*.nation.*", privilege("nationkey", SELECT_COLUMN));
```
Comment (findepi) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: + `assertAccessDenied("select * from nation")`

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java (hunk 10 above)
Comment (rschlussel2) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java]: That was my first thought, but FieldReference doesn't go through that path, so I would have to duplicate the logic for visitFieldReference.

File: presto-plugin-toolkit/src/main/java/com/facebook/presto/plugin/base/security/ReadOnlyAccessControl.java (hunk 11 above)
Comment (rschlussel2) [reply presto-plugin-toolkit/src/main/java/com/facebook/presto/plugin/base/security/ReadOnlyAccessControl.java]: The other `checkCanCreateViewWithSelectFrom`s were next to each other and rather than their `checkCanSelectFrom` counterparts.  That seemed as good an organizing principle.

**Comment by findepi** [comment]
Do you need this logic for FieldReference?
I thought this is needed for post-analyzer query processing

**Comment by rschlussel2** [comment]
@findepi I'm not sure I understand your point, so I'm just going to outline the information I'm working with and you can tell me if I didn't address what you meant.

1. The logic I added is used to collect the table-column mappings for `Analysis` so that it can be used in the `StatementAnalyzer` to check column-level permissions. 
2. FieldReferences are among the kinds of expressions that the `StatementAnalyzer` has and needs to analyze (e.g. the columns from a `select *` get represented as FieldReferences). 
3. `visitFieldReference()` calls `handleResolvedField(.., FieldId, ...)` directly rather than going through`handleResolvedField(.., ResolvedField, ...)`.

**Comment by findepi** [comment]
I know (i think so) `FieldReference`s are used in physical plan (channels) and i think we do some expression analysis on worker nodes -- this is why i thought skipping this would be ok.

*However*,  i see handling of `select * from t` now: https://github.com/prestodb/presto/blob/6f010a60e66ea0189e1c02817b4dfb6ecae475ff/presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java#L1806-L1809

`FieldReference` are created here. Why? Maybe because not all fields have a name (`com.facebook.presto.sql.analyzer.Field#name` is optional)? But then again, `com.facebook.presto.spi.ColumnMetadata#name` is not optional (uff), and your auth checks work with `Set<String> columns`. Maybe we could the above handling in `StatementAnalyzer` to use names (where possible)?

**Comment by rschlussel2** [comment]
It's not that the columns don't have names, it's that they aren't named in the queries.  So what we start with are the fields of the table, and from there it doesn't make sense to get the column names just to resolve it back to a FieldIndex.   We do the same for something like `ORDER BY 1` (see https://github.com/prestodb/presto/blob/6f010a60e66ea0189e1c02817b4dfb6ecae475ff/presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java#L2124).

**Comment by martint** [comment]
We may want to think about how to model SELECT permissions on a table by relying on the SELECT permissions for the table columns. At the end of the day, the SQL spec doesn't really have a concept of SELECT permission for a table. It's derived from the SELECT permissions for the columns of the table:

```
1) If a <table primary> TP simply contains a <table or query name> that simply 
     contains a <table name> TN, then:
  a) Let T be the table identified by TN.
    [...]
    B) Otherwise, the current privileges shall include SELECT on at least one column of T.
```

File: presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorAccessControl.java [hunk 13]
```diff
@@ -258,6 +259,11 @@ default void checkCanCreateViewWithSelectFromView(ConnectorTransactionHandle tra
         denyCreateViewWithSelect(viewName.toString());
     }
 
+    default void checkCanCreateViewWithSelectFromColumns(ConnectorTransactionHandle transactionHandle, Identity identity, SchemaTableName tableName, Set<String> columnNames)
```
Comment (electrum) [inline presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorAccessControl.java]: Add Javadoc for this, like the other methods.
```
* Check if identity is allowed to create a view that selects from the specified columns in this catalog.

File: presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorAccessControl.java [hunk 14]
```diff
@@ -287,4 +293,14 @@ default void checkCanRevokeTablePrivilege(ConnectorTransactionHandle transaction
     {
         denyRevokeTablePrivilege(privilege.toString(), tableName.toString());
     }
+
+    /**
+     * Check if identity is allowed to select from the specified columns in a table.
+     *
+     * @throws com.facebook.presto.spi.security.AccessDeniedException if not allowed
+     */
+    default void checkCanSelectFromColumns(ConnectorTransactionHandle transactionHandle, Identity identity, SchemaTableName tableName, Set<String> columnNames)
```
Comment (electrum) [inline presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorAccessControl.java]: Should we put this next to the table permissions, rather than adding it at the end? Keeping similar methods next to each other is helpful.

Same question for other interfaces and implementations.

File: presto-main/src/main/java/com/facebook/presto/security/ViewAccessControl.java [hunk 15]
```diff
@@ -52,4 +54,10 @@ public void checkCanCreateViewWithSelectFromView(TransactionId transactionId, Id
     {
         delegate.checkCanCreateViewWithSelectFromView(transactionId, identity, viewName);
     }
+
+    @Override
+    public void checkCanSelectFromColumns(TransactionId transactionId, Identity identity, QualifiedObjectName tableName, Set<String> columnNames)
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/security/ViewAccessControl.java]: Let's move this above, so that all the `checkCanSelect` methods are together.

File: presto-main/src/main/java/com/facebook/presto/security/AccessControl.java [hunk 16]
```diff
@@ -193,6 +193,13 @@
      */
     void checkCanCreateViewWithSelectFromView(TransactionId transactionId, Identity identity, QualifiedObjectName viewName);
 
+    /**
+     * Check if identity is allowed to create a view that selects from the specified view.
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/security/AccessControl.java]: This should be `from the specified columns`

File: presto-main/src/main/java/com/facebook/presto/security/AccessControl.java [hunk 17]
```diff
@@ -220,4 +227,11 @@
      * @throws com.facebook.presto.spi.security.AccessDeniedException if not allowed
      */
     void checkCanSetCatalogSessionProperty(TransactionId transactionId, Identity identity, String catalogName, String propertyName);
+
+    /**
+     * Check if identity is allowed to select from the specified column.
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/security/AccessControl.java]: `column` -> `columns`

File: presto-main/src/main/java/com/facebook/presto/testing/TestingAccessControlManager.java [hunk 18]
```diff
@@ -336,7 +362,7 @@ private boolean shouldDenyPrivilege(String userName, String entityName, TestingP
         ADD_COLUMN, DROP_COLUMN, RENAME_COLUMN,
         CREATE_VIEW, DROP_VIEW, SELECT_VIEW,
         CREATE_VIEW_WITH_SELECT_TABLE, CREATE_VIEW_WITH_SELECT_VIEW,
-        SET_SESSION
+        SET_SESSION, SELECT_COLUMN
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/testing/TestingAccessControlManager.java]: This isn't related to `SET_SESSION`. Let's add it on the line with other permissions (`ADD_COLUMN`, etc.)

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 19]
```diff
@@ -683,6 +683,11 @@ protected Scope visitQuery(Query node, Optional<Scope> scope)
                     .withParent(withScope)
                     .withRelationType(RelationId.of(node), queryBodyScope.getRelationType())
                     .build();
+
+            // check column access permissions for each table
+            analysis.getTableColumnMap().asMap()
+                    .forEach((tableName, columns) -> accessControl.checkCanSelectFromColumns(session.getRequiredTransactionId(), session.getIdentity(), tableName, ImmutableSet.copyOf(columns)));
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: This line is quite long. Let's wrap here for readability.

File: presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorAccessControl.java [hunk 20]
```diff
@@ -287,4 +293,14 @@ default void checkCanRevokeTablePrivilege(ConnectorTransactionHandle transaction
     {
         denyRevokeTablePrivilege(privilege.toString(), tableName.toString());
     }
+
+    /**
+     * Check if identity is allowed to select from the specified columns in a table.
```
Comment (electrum) [inline presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorAccessControl.java]: Technically, this should be `columns in a relation` as this applies to both tables and views.

File: presto-spi/src/main/java/com/facebook/presto/spi/security/AccessDeniedException.java [hunk 21]
```diff
@@ -273,6 +274,16 @@ public static void denySetCatalogSessionProperty(String propertyName)
...
     }
 
+    public static void denySelectColumns(String tableName, Collection<String> columnNames)
+    {
+        denySelectColumns(tableName, columnNames, null);
+    }
+
+    public static void denySelectColumns(String tableName, Collection<String> columnNames, String extraInfo)
+    {
+        throw new AccessDeniedException(format("Cannot select from columns %s in table %s%s", columnNames, tableName, formatExtraInfo(extraInfo)));
```
Comment (electrum) [inline presto-spi/src/main/java/com/facebook/presto/spi/security/AccessDeniedException.java]: This can also be a view: `table` -> `relation`

Or we can say `in table or view %s` (might be better to avoid the term "relation" in user error messages)

File: presto-spi/src/main/java/com/facebook/presto/spi/security/SystemAccessControl.java [hunk 22]
```diff
@@ -292,6 +293,16 @@ default void checkCanCreateViewWithSelectFromView(Identity identity, CatalogSche
         denyCreateViewWithSelect(view.toString());
     }
 
+    /**
+     * Check if identity is allowed to create a view that selects from the specified columns in a table.
```
Comment (electrum) [inline presto-spi/src/main/java/com/facebook/presto/spi/security/SystemAccessControl.java]: `table` -> `relation`

File: presto-spi/src/main/java/com/facebook/presto/spi/security/SystemAccessControl.java [hunk 23]
```diff
@@ -321,4 +332,14 @@ default void checkCanRevokeTablePrivilege(Identity identity, Privilege privilege
     {
         denyRevokeTablePrivilege(privilege.toString(), table.toString());
     }
+
+    /**
+     * Check if identity is allowed to select from the specified columns in a table.
```
Comment (electrum) [inline presto-spi/src/main/java/com/facebook/presto/spi/security/SystemAccessControl.java]: `table` -> `relation`

**Comment by rschlussel2** [comment]
@martint From here it would make sense to me to get rid of checkCanSelectFromTable entirely, and connectors that don't support column level security can implement just a table check in checkCanSelectFromColumns.  The only addition we would need to make would be to submit a request with an empty column list for queries that don't select any columns, such as `select count(*) from foo`

**Comment by electrum** [comment]
@rschlussel2 I like that approach. It will also remove the distinction between tables and views, which I'm not sure was useful in the first place. Thoughts @dain?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java [hunk 24]
```diff
@@ -284,6 +288,11 @@ private Type analyze(Expression expression, Scope baseScope, Context context)
         return unmodifiableSet(windowFunctions);
     }
 
+    public Multimap<QualifiedObjectName, String> getTableColumnMap()
```
[the remark this answers is not in the document]
Comment (rschlussel2) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java]: sure

File: presto-main/src/main/java/com/facebook/presto/testing/TestingAccessControlManager.java (hunk 18 above)
Comment (electrum) [reply presto-main/src/main/java/com/facebook/presto/testing/TestingAccessControlManager.java]: Please move this above to be with the other column privileges.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java [hunk 25]
```diff
@@ -89,6 +89,11 @@ private Scope(
         return Optional.empty();
     }
 
+    public Optional<Scope> getParent()
+    {
+        return parent;
+    }
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java]: When one should call `.getParent()` and when `. getOuterQueryParent()`?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java [hunk 26]
```diff
@@ -400,7 +400,7 @@ private Type handleResolvedField(Expression node, FieldId fieldId, Type resolved
                 }
             }
 
-            Field field = baseScope.getRelationType().getFieldByIndex(fieldId.getFieldIndex());
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java]: I think there is better approach. You can change signature of this method to:
```
handleResolvedField(Expression node, FieldId fieldId, Field field, StackableAstVisitorContext<Context> context)
```
and then modify callers as:
```
private Type handleResolvedField(Expression node, ResolvedField resolvedField, StackableAstVisitorContext<Context> context)
{
    return handleResolvedField(node, FieldId.from(resolvedField), resolvedField.getField(), context);
}
```
and
```
public Type visitFieldReference(FieldReference node, StackableAstVisitorContext<Context> context)
{
    Field field = baseScope.getRelationType().getFieldByIndex(node.getFieldIndex());
    return handleResolvedField(node, new FieldId(baseScope.getRelationId(), node.getFieldIndex()), field, context);
}
```

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java [hunk 27]
```diff
@@ -410,6 +410,17 @@ private Type handleResolvedField(Expression node, FieldId fieldId, Type resolved
             return setExpressionType(node, resolvedType);
         }
 
+        private RelationType getSourceRelation(Scope scope, RelationId relationId)
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java]: I don't think it's needed, see my comment on https://github.com/prestodb/presto/pull/10375/files/098ddf08b4bfb2061fc3d855a307f5d14098ff5b..9949b16ce62d2a0c9b09291e3a2546cb3b01ef6b#diff-de0c3bd1446c0ac0bf16a9e1c91b20f5L403

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java [hunk 28]
```diff
@@ -89,6 +89,11 @@ private Scope(
         return Optional.empty();
     }
 
+    public Optional<Scope> getParent()
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java]: I don't think it's needed, see my comment on: I don't think it's needed, see my comment on https://github.com/prestodb/presto/pull/10375/files/098ddf08b4bfb2061fc3d855a307f5d14098ff5b..9949b16ce62d2a0c9b09291e3a2546cb3b01ef6b#diff-de0c3bd1446c0ac0bf16a9e1c91b20f5L403

**Comment by sopel39** [comment]
@rschlussel2 Please address my https://github.com/prestodb/presto/pull/10375#discussion_r183315842 comment

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java (hunk 25 above)
Comment (rschlussel2) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java]: when we resolve a field, we recurse through the parents independent of whether it crosses a query boundary and stop when we find a match, so we need to do the same process to get the field that goes with the fieldId.  If you want specifically to get the inner or outer query you would call getInnerQueryParent() or getOuterQueryParent()

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java (hunk 26 above)
Comment (rschlussel2) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java]: Thanks.  I did a variation.

**Comment by rschlussel2** [comment]
@sopel39 can I merge this now?

**Comment by sopel39** [comment]
@rschlussel2 looks good

**Comment by rschlussel2** [comment]
Thanks @sopel39.  I squashed in the fixups and I'll merge after the release goes out.
